src/test/java/org/team199/wpiws/devices/SimDeviceSimTest.java

```diff
@@ -0,0 +1,471 @@
+
+            order = inOrder(notifiedCallbackAfterFirstValueInit);
+            order.verify(notifiedCallbackAfterFirstValueInit)
+                    .callback(eq("value1"), eq("data1"));
+            order.verify(notifiedCallbackAfterFirstValueInit)
+                    .callback(eq("value2"), eq("data2"));
+            verifyNoMoreInteractions(notifiedCallbackAfterFirstValueInit);
+
+            order = inOrder(nonNotifiedCallbackAfterFirstValueInit);
+            order.verify(nonNotifiedCallbackAfterFirstValueInit)
+                    .callback(eq("value2"), eq("data2"));
+            verifyNoMoreInteractions(nonNotifiedCallbackAfterFirstValueInit);
+
+
+        } finally {
+            callbacks.forEach(sim::cancelValueCreatedCallback);
+        }
+    }
+
+    /**
+     * Creates a device name that will not conflict with devices from other tests. Devices names for
+     * the same test case will always be equal.
+     *
+     * @param testMethodName the name of the test being run
+     * @return a device name unique to the given test
+     */
+    private static String getDeviceName(String testMethodName) {
+        return "SimDeviceSimTest.%s".formatted(testMethodName);
+    }
+
+    @RunWith(Parameterized.class)
+    public static class DataTests<VarType> {
+
+        @Parameters(name = "{index}: {0}")
+        public static Object[] testCases() {
+            return Arrays.stream(new Object[][] {
+                    createTestCase("BooleanValue", SimDeviceSim::set, false,
+                            true),
+                    createTestCase("DoubleValue", SimDeviceSim::set, 0.0, 0.1),
+                    createTestCase("StringValue", SimDeviceSim::set, "",
+                            "data")})
+                    .flatMap(Arrays::stream).toArray();
+        }
+
+        /**
+         * Uses the given information to construct the necessary parameters to run a test. This
+         * method creates three test cases that simulate values that are a a robot input, a robot
+         * output, and a bidirectional value, respectively.
+         *
+         * @param <VarType> the type of variable being tested
+         * @param valueName a name for the variable. This will be included in the test name. This is
+         *        not a WS name.
+         * @param setterFunction the specific SimDeviceSim function to use to set the value of the
+         *        variable
+         * @param defaultValue the "default" value of the variable. For SimDeviceSim, get() always
+         *        returns `null` by default, but we still need a specific value for testing.
+         *        Ensuring that this is "falsy" will prevent any conflicts.
+         * @param alternateValue an alternative example of this value type that can be used for
+         *        testing
+         * @return an Object array containing the generated test cases
+         */
+        private static <VarType> Object[][] createTestCase(String valueName,
```

Comment:
Rename to `createTestCases` and update first sentence of javadoc.

Reply:
Done!